Add DAW Controller support for various Arturia controllers

soyersoyer opened the pull request:
This feature can be enabled in the minidexed.ini with:
```
DAWControllerEnabled=1
```

## Features

### Controller Display, Main Encoder
The controller display in DAW mode shows the MiniDexed menu.
With the main encoder you can navigate in the menu.
The Home is the shift+Main click

Supported:
- MiniLab 3 (tested)
- KeyLab mkII (tested)
- Keylab Essential
- Keylab Essential 3

### Energy efficiency
This version uses a little less power so it runs cooler thanks to #3.

### Extra MIDI CC
- Sostenuto
- Hold2
- Portamento

### MiniLab 3 features:

#### Encoders
By holding down shift will bring up an overlay menu, where you can see and adjust what the encoders do.
You can use the main encoder to select the current encoding page.
Short pressing the shift will show the actual values.
Holding down the shift will goes back to the normal menu.
This overlay menu is context-aware, it is different if you are in the effect menu, or in a TG menu.
If you are in a TG menu, it only affects the selected TG.
If you are in the main menu, it affects the TGs on the first TG's channel.

- Main Overlay 1 (default)
Cutoff, Resonance, ReverbSend, None
PortamentoTime, Program, Volume, None
Faders: ChG 1 Vol, ChG 2 Vol, ChG 3 Vol, ChG 4 Vol

- Effect Overlay 1
CompressorEnable, ReverbEnable, ReverbSize, ReverbHighDamp
ReverbLowDamp, ReverbLowPass, ReverbDiffusion, ReverbLevel
Faders: ReverbSize, ReverbHighDamp, ReverbLowDamp, ReverbLevel

- Effect Overlay 2
LimiterEnable, LimiterPreGain, LimiterAttack, LimiterRelease
LimiterThresh, LimiterRatio, LimiterHPFilter, None
Faders: LimiterPreGain, LimiterAttack, LimiterRelease, LimiterThresh

- Main Overlay 2
Pan1, Pan2, Pan3, Pan4
Det1, Det2, Det3, Det4
Faders: TG1 Vol, TG2 Vol, TG3 Vol, TG4 Vol

- Main Overlay 3
Pan5, Pan6, Pan7, Pan8
Det5, Det6, Det7, Det8
Faders: TG5 Vol, TG6 Vol, TG7 Vol, TG8 Vol

- TG Overlay 1
Cutoff, Resonance, ReverbSend, MasterTune
PortamentoTime, Program, Volume, Pan
Faders: Cutoff, Resonance, ReverbSend, Volume

- TG Overlay 2
MIDIChannel, None, None, PitchBendRange
PortamentoGlissando, MonoMode, None, PitchBendStep

- TG Overlay 3
CompressorEnable, CompressorPreGain, CompressorAttack, CompressorRelease
CompressorThresh, CompressorRatio, None, None
Faders: CompressorPreGain, CompressorAttack, CompressorRelease, CompressorThresh

- TG Overlay 4
MWRange, MWPitch, FCRange, FCPitch
MWEGBias, MWAmplitude, FCEGBias, FCAmplitude

- TG Overlay 5
BCRange, BCPitch, ATRange, ATPitch
BCEGBias, BCAmplitude, ATEGBias, ATAmplitude

- Voice Overlay 1
ALGORITHM, FEEDBACK, TRANSPOSE, None,
None, None, None, None

- Voice Overlay 2
PITCH_EG_R1, PITCH_EG_R2, PITCH_EG_R3, PITCH_EG_R4
PITCH_EG_L1, PITCH_EG_L2, PITCH_EG_L3, PITCH_EG_L4
Faders: PITCH_EG_L1, PITCH_EG_L2, PITCH_EG_L3, PITCH_EG_L4

- Voice Overlay 3
OSC_KEY_SYNC, LFO_SPEED, LFO_DELAY, LFO_PITCH_MOD_DEP
LFO_SYNC, LFO_WAVE, LFO_PITCH_MOD_SENS, LFO_AMP_MOD_DEP

- OP Overlay 1
OP_OUTPUT_LEV, OP_FREQ_COARSE, OP_FREQ_FINE, OP_OSC_DETUNE,
OP_OSC_MODE, OP_ENABLE, None, None

- OP Overlay 2
OP_EG_R1, OP_EG_R2, P_EG_R3, OP_EG_R4
OP_EG_L1, OP_EG_L2, OP_EG_L3, OP_EG_L4,
Faders: OP_EG_L1, OP_EG_L2, OP_EG_L3, OP_EG_L4,

- OP Overlay 3
OP_LEV_SCL_BRK_PT, OP_SCL_LEFT_DEPTH, OP_SCL_RGHT_DEPTH, OP_AMP_MOD_SENS
OP_OSC_RATE_SCALE, OP_SCL_LEFT_CURVE, OP_SCL_RGHT_CURVE, OP_KEY_VEL_SENS

In the Main overlay you can reach the TG and voice overlays also.
In the voice overlay there is a page for every OP control also, with the 6 operator and with an encoder which sets all.

#### Pads on Bank A
- Mono/Poly mode
- Portamento enable/disable, hard press to enable/disable Portamento Glissando
- Sostenuto
- Sustain
- All Sound Off
- Hold2
- None
- Channel AfterTouch Pad

#### Pads on Bank B
- Short press a TG will enable/disable that TG
- Long press will enable/disable all the TG on the TG's channel
- Hard press will enable only that TG (it uses the pad's AT messsages)


### KeyLab mkII features

#### Faders
- The faders adjust the volume of the MIDI channels of the current performance

#### DAW buttons
- Mono/Poly mode
- Portamento enable/disable, hard press to enable/disable Portamento Glissando
- Sostenuto
- Sustain
- Hold2

#### Track select buttons
- Short press a TG will enable/disable that TG
- Long press will enable only that TG

soyersoyer committed:
add DAW Controller support for various Arturia controllers

MiniLab 3
Keylab Essential
KeyLab Essential mk3
KeyLab mkII

based on https://github.com/PrzemekBarski/arturia-keylab-essential-mk3-programming-guide

Tested on a Arturia MiniLab 3 and KeyLab mkII

Keylab Essential and Keylab Essential mk3 is not tested

soyersoyer committed:
add MIDI Button actions

soyersoyer committed:
CUIBUttons: add dec and inc trigger modes

soyersoyer committed:
Add TGParameterEnabled

soyersoyer committed:
add synchronization for multcore operations

soyersoyer committed:
suspend cores while waiting

It works with WaitForEvent, without disabling IRQs.

Before this the WaitForInterrupt was used, and these findigs were useful:

Before checking the atomic variable, disable IRQs to avoid missing the wake-up event.

https://marc.info/?l=linux-arm-kernel&m=114684008423863&w=2
https://electronics.stackexchange.com/questions/12601/best-pattern-for-wfi-wait-for-interrupt-on-cortex-arm-microcontrolers

add dsb before wfi

https://github.com/torvalds/linux/commit/8553cb67d2318db327071018fc81084cbabccc46
https://github.com/torvalds/linux/blob/master/arch/arm/mm/proc-v7.S
https://github.com/torvalds/linux/blob/master/arch/arm64/kernel/idle.c
https://github.com/torvalds/linux/commit/9cce7a435f89c9e60f244d44da2cf1cf4ed094ac

soyersoyer commented:
News:
- Better KeyLab mkII support
- Fewer unnecessary display/encoder/pad color updates
- Uses less power

~It turned out that it doesn't work well with the MiniLab 3 firmware v1.2.0. Please use version v1.1.1.~

Spacepadrille commented:
Hello once again thank you for your work. I enjoy a lot my MiniDexed since your arturia mod is there. About the Minilab3 firmware update, can you tell us more ? Do you think it will be solved or do we have to stay with previous version forever ? 
Have a nice day

soyersoyer commented:
> Hello once again thank you for your work. I enjoy a lot my MiniDexed since your arturia mod is there.

So am I (:

The new version has this [feature](https://www.arturia.com/api/resources/release_notes?resource_id=24199):

> Automated controls are no longer shown on the MiniLab 3 screen unless they are actively adjusted

This blocks all screen Info (Pad, Fader, Encoder) updates if one of the encoders or faders or pads is not in use. And whatever decides what is used is not good yet for MiniDexed.

I have 2 problems:
 - If you long press a pad on Bank B, it won't show on the display what happening after the timeout. (I reported this: https://forum.arturia.com/t/minilab-3-firmware-v1-2-0/6064)
 -  If you press a pad in Bank A very lightly (except Portamento and AT), the display does not show what is happening. I found this particularly annoying. This may be because I send the screen update immediately after the NOTE_ON message, not after the NOTE_OFF, so I may be able to fix this.

Okay, the latter has been fixed. Thanks for your question :)

I'll update the description to say that v1.2.0 is just a little bad :)

soyersoyer commented:
I also fixed the former one by displaying it without the image of the pad.

soyersoyer commented:
News:
 - MiniLab3 Main encoder pages: add Pan/Detune pages (TG 1-4, TG 5-8) with Volume faders
 - MiniLab3 TG encoder pages: add Cutoff/Resonance/ReverbSend/Volume faders to the first page
 - MiniLab3 Voice encoder pages: add Pitch EG Level faders
 - MiniLab3 OP encoder pages: add EG level faders

soyersoyer commented:
- Add Effect page to the Main overlay
- Add volume encoder to the Main page 1

soyersoyer commented:
The latest version is now available for download from the releases and there is a description in the readme.
I'll leave this open to write the news here and for feedback.

soyersoyer committed:
mixer: fix pan channel order

0.0 full left
1.0 full right

soyersoyer committed:
hdmi sound: remove default channel swap
as it's fixed in the pan

soyersoyer committed:
effect_mixer: add ramping and zero cross detection

Ramp the scale parameter and change it only at zero crossings
to minimize pops and clicks.

soyersoyer committed:
ramp the MasterVolume at zero crossings

soyersoyer committed:
scale_zip: use f32x4x2_t

soyersoyer committed:
CMiniDexed: add MasterVolume global parameter

Handle MasterVolume as other global parameters to
simplify code.

soyersoyer commented:
- Added Master Volume to Minilab3 Main Overlay 1 Fader4
- Added Master Volume to Keylab2 Fader9

soyersoyer committed:
midi: remove the BANK typo from MIDI_CC_BANK_SUSTAIN

soyersoyer committed:
CMiniDexed: member initialize pointers to nullptr

soyersoyer committed:
CUIMenu: pass nWidth parameter to ToVolume functions

Enable LCDColumn wide Volume and Pan adjustments and print value.

soyersoyer committed:
CUIMenu: do not use the full std namespace

The using namespace can cause confusions
https://stackoverflow.com/questions/1452721/whats-the-problem-with-using-namespace-std

soyersoyer committed:
Add option to configure default screen

Add DefaultScreen option to the minidexed.ini
0=Default
1=Performance Load

Open the Performance>Load menu if Performance Load is set

soyersoyer committed:
CMiniDexed: show Network Ready + IP instead of TG1

TG1 is misleading if the menu is not on it.

Show only for for 3 seconds.

soyersoyer committed:
UI: Set volume and pan increments to 1 if Encoder is enabled

soyersoyer committed:
PerfTimer: add lastMicros

Add ability to write the last measured time.

soyersoyer committed:
PerfTimer: add CPU information

soyersoyer committed:
kernel: RegisterSystemThrottledHandler

soyersoyer committed:
move compressor into the TG menu

soyersoyer committed:
add compressor parameters

probonopd committed:
Merge pull request #966 from soyersoyer/nobank

midi: remove the BANK typo from MIDI_CC_BANK_SUSTAIN

soyersoyer committed:
update Synth_Dexed to a02b5c0bf2

soyersoyer committed:
config: add LogThrottling option

Add an option to enable logging of throttling events.

soyersoyer committed:
kernel: register systemThrottledHandler if the LogThrottling config option is enabled

Logging can negativel impact the sound path.
This can also occur in otherwise functional setups, so it's better not to have it enabled in all cases.

soyersoyer committed:
performanceconfig: fix warnings from newer compilers

calling std::string::operator= with nullptr is invalid, so clear it instead.

soyersoyer committed:
uimenu: use math.h instead of cmath

the cmath include doesn't work with arm_math.h with gcc13.3

soyersoyer committed:
remove the unused CStdlibAppNetwork

this suppress warning about the incompatible overloaded Initialize() function

soyersoyer committed:
gitignore: add the new arm-gnu-* toolchain prefix

soyersoyer committed:
Makefile: add $(OBJS) and $(DEPS) to EXTRACLEAN and use the existing clean

soyersoyer committed:
gh build: use gcc13.3

soyersoyer committed:
remove the obsolete USE_FX from Synth_Dexed Makefile

this has already been removed from Synth_dexed

https://codeberg.org/dcoredump/Synth_Dexed/commit/b75b95986ac1f116809f1445e05e557f4ae79384

soyersoyer committed:
add limiter

soyersoyer committed:
config: return CIPAddress as const reference

returning by value calls
CIPAddress::CIPAddress (const CIPAddress &rAddress)
which assert (rAddress.m_bValid) so any unset ip will fail in debug builds

soyersoyer committed:
config: set ip addresses only if they are valid

soyersoyer committed:
InitNetwork: check IP addresses properly

soyersoyer committed:
gh build: create DreamDexed zips

soyersoyer committed:
README: update to DreamDexed

[ci skip]

soyersoyer committed:
README: update wiki links

[ci skip]

soyersoyer committed:
CMiniDexed: fix crash when a performance file has more TGs than m_nToneGenerators

soyersoyer committed:
CMiniDexed: fix LoadPerformanceParameters indent

soyersoyer committed:
README: update to DreamDexed

soyersoyer committed:
add 3 Band EQ per tg

soyersoyer committed:
3BandEQ: remove the unneeded denormal fix

Why it was there:
https://ldesoras.fr/doc/articles/denormal-en.pdf

but we don't need because Circle enables the flush to zero mode
https://github.com/rsta2/circle/blob/927941cec722f530a23d0f636f03480dd9972056/lib/sysinit.cpp#L224

soyersoyer committed:
CMiniDexed: zero initialize m_pTG

soyersoyer committed:
add master EQ

soyersoyer committed:
put Compressor after TG EQ

Do not use the compressor in Dexed instances,
use the same one, but instantiate in MiniDexed

soyersoyer committed:
sendfx

soyersoyer committed:
add chorus

soyersoyer committed:
add lpf effect

soyersoyer committed:
add delay effect

soyersoyer committed:
add fx chain

soyersoyer committed:
README: add 3-band EQ to feature list

[ci skip]

soyersoyer committed:
update default reverb parameters

based on suggestions from Banana71

soyersoyer committed:
effect_mixer: use x^2 for gain

modify the FxSends also to use values matching to x^2

soyersoyer committed:
CMiniDexed: SetParameter: save the parameter after limiting it to the limits

soyersoyer committed:
reverb: replace disable/enable with a dry/wet mixer

soyersoyer committed:
mixer: doAddMix(): skip processing if multiplier is 0

soyersoyer committed:
add MixerDryLevel control

soyersoyer committed:
Menu: show mix values as Dry:Wet format

soyersoyer committed:
Update to circle-stdlib v17 (circle50)

soyersoyer committed:
update toolchain to ARM GNU 14.3.Rel1

soyersoyer committed:
Add AudioEffectCloudSeed

soyersoyer committed:
use the c++ version of c includes

soyersoyer committed:
CUIMenu: add note limits and shift

soyersoyer committed:
Add Compressor Makeup gain

soyersoyer committed:
CUIMenu: add OnSelect for EditTGParameter2

soyersoyer committed:
CMiniDexed: add lastKeyDown getting functionality

soyersoyer committed:
CUIMenu: call InputKeyDown for note limits

soyersoyer committed:
CUIMenu: call InputShiftKeyDown for note shift

Banana71 committed:
Update README.md

Cloud Seed MIT license

soyersoyer committed:
reorder compressor and limiter fields

soyersoyer committed:
Volume ramping before and after performance change

soyersoyer committed:
AudioEffect3BandEQ: add resetState() function

soyersoyer committed:
Add WaitForEvent for Core 1 also to reduce power consumpiton

This reduces the power consumption when used in fast=true mode

soyersoyer committed:
effect_mixer: use simple slow ramping instead of zero cross detection

Some performances have a DC component while silent, which prevents ramping.

soyersoyer committed:
reverb: add explicit reset

soyersoyer committed:
AudioMixer: simplify scale_ramp_f32

soyersoyer committed:
Synth_Dexed: update to e3f524c8b (PluginFx: use small ramping instead of zero cross detection)

soyersoyer committed:
Add TG-Links

Now every TG has a TG-Link parameter.
default: -
Possible values: - A B C D
If you modify a parameter whose TG belongs to a TG-Link,
the parameter of the TGs belonging to the group will also be modified.

Except Pan, MasterTune and TG-Link parameters.

soyersoyer committed:
CUIMenu: TGShortcutHandler switch between menu items

Now we have TG-Link so switching between TG-s is not necessary anymore.

soyersoyer committed:
CUIMenu: put link names after TG names

soyersoyer committed:
CUIMenu: add GlobalShortcutHandler

soyersoyer committed:
README: add tg-link and noiseless performance

soyersoyer committed:
CUIMenu: save current selection in GlobalShortcutHandler

soyersoyer committed:
rename Limiter to MasterCompressor

soyersoyer committed:
add SendFX slots

soyersoyer committed:
CUserInterface::DisplayWrite: limit the lines to LCDColumns length

soyersoyer committed:
performanceconfig: set FX2Send[n] = 0 by default

soyersoyer committed:
performanceconfig: use the MasterCompressor as Minidexed performance Compressor

soyersoyer committed:
UIMenu: add Mixer menu

soyersoyer committed:
update circle-stdlib to v17.2

soyersoyer committed:
status menu

soyersoyer committed:
README: update status

soyersoyer committed:
PerformanceConfig: use MiniDexed compressor settings in MiniDexed compatibility mode

soyersoyer committed:
CMiniDexed: use the compressor ratio value 31 as infinity

soyersoyer committed:
CMiniDexed: set max compressor release parameter to 2000 ms

soyersoyer committed:
CMiniDexed: set min compressor attack/release time to 1 ms

soyersoyer committed:
PerformanceConfig: refine compressor settings in MiniDexed compatibility mode

soyersoyer committed:
CUIMenu: ShowCPU*: retrieve values from CStatus

soyersoyer committed:
CUIMenu: show TG-Link in Voice menu

soyersoyer committed:
effect: move functions and arrays into the appropriate files

soyersoyer committed:
effect_cloudseed2: move presets to cpp

soyersoyer committed:
effect_cloudseed2: convert to tabs

soyersoyer committed:
add AudioEffectCompressor

soyersoyer committed:
rename AudioEffect3BandEQ to AudioEffect3BandEQMono

soyersoyer committed:
effect_3bandeqmono: convert to tabs

soyersoyer committed:
effect_3bandeqmono: use float samplerate

soyersoyer committed:
add AudioEffect3BandEQ

soyersoyer committed:
use FXChain as master FX

soyersoyer committed:
effect_chain: scale only if necessary

soyersoyer committed:
effect_ykchorus: add bypass

soyersoyer committed:
effect_dreamdelay: add bypass

soyersoyer committed:
effect_platereverb: add bypass

soyersoyer committed:
effect_cloudseed2: add bypass

soyersoyer committed:
Update CloudSeedCore

soyersoyer committed:
add FXBypass option to bypass all FX chains

Banana71 committed:
Bank 004 Mirage DreamDexed Performances (#24)

* Bank 004 Mirage DreamDexed Performances

* Copy performance files to sdcard

soyersoyer committed:
build: copy the DreamDexed Bank 004_Mirage.pdf to the root of the sdcard

Banana71 committed:
Remove network ready display and timer start

Removed UI display write and timer start for network readiness.

soyersoyer committed:
uimenu: showdirect

soyersoyer committed:
status menu: add Network IP

soyersoyer committed:
CMIDIPin: convert the MIDI value to LOW/HIGH in CUIButton

simplify the CMIDIPin in order to easier to handle dec/inc values

soyersoyer committed:
SetVoiceParameter: reverse the nOP only once

soyersoyer committed:
Status: read the actual IP address from the network subsystem

not just from the config

soyersoyer committed:
CUIButtons: remove the unused timeout members

soyersoyer committed:
CUIButtons: add debounce time for Relative MIDI messages

It's needed for MPK MINI mk3

soyersoyer committed:
README: add relative mode encoders

soyersoyer committed:
Status: add version

soyersoyer committed:
gh build: remove startup version message

soyersoyer committed:
gh build: use commiter date instead of the actual date

This is the same what used in the Makefile

soyersoyer committed:
gh build: use checkout@v6 and pull the PR head instead of the merge commit

soyersoyer committed:
Makefile: use shorter date in version

soyersoyer committed:
gh build: use shorter date in version
